File: apps/mcp-server/src/tools/identity.ts
/**
 * MCP tools for the `agent-identity` module.
 *
 * Mirrors `apps/agent-identity/src/mcp/tools.ts`:
 *   - colber_identity_register
 *   - colber_identity_resolve
 *   - colber_identity_verify
 *
 * Tool names use the `colber_<module>_<verb>` convention (snake-case verb,
 * underscore-separated) to avoid collisions with other MCP servers a user
 * might have installed.
 */

import { z } from 'zod';

import type { ToolRegistry } from './registry.js';
import type { ColberClient } from '@colber/sdk';

export const registerIdentityTools = (registry: ToolRegistry, sdk: ColberClient): void => {
  registry.register({
    name: 'colber_identity_register',
    description:
      '[Colber] Register a new agent identity from an Ed25519 public key. Returns the derived did:key, the internal agent UUID, and the registration timestamp.',
    inputSchema: z.object({
      publicKey: z.string().min(1).describe('Ed25519 public key, base64 (32 raw bytes)'),
      ownerOperatorId: z.string().min(1).max(128),
    }),
    handler: (input) =>
      sdk.identity.register({
        publicKey: input.publicKey,
        ownerOperatorId: input.ownerOperatorId,
      }),
  });

  registry.register({
    name: 'colber_identity_resolve',
    description: '[Colber] Resolve a DID to its agent record.',
    inputSchema: z.object({
      did: z.string().min(1),
    }),
    handler: (input) => sdk.identity.resolve(input.did),
  });

  registry.register({
    name: 'colber_identity_verify',
    description:
      '[Colber] Verify a signature against the public key bound to a DID. Returns { valid: true|false }.',
    inputSchema: z.object({
      did: z.string().min(1),
      message: z.string().min(1).describe('base64-encoded message bytes'),
      signature: z.string().min(1).describe('base64-encoded signature bytes'),
    }),
    handler: (input) =>
      sdk.identity.verify({
        did: input.did,
        message: input.message,
        signature: input.signature,
      }),
  });
};

File: apps/colber-langchain/src/colber_langchain/tools/identity.py
"""Identity service tools — register + resolve agents."""

from __future__ import annotations

from typing import Any, ClassVar

from pydantic import BaseModel, Field

from ._base import ColberToolBase


class _RegisterArgs(BaseModel):
    public_key: str = Field(
        description=(
            "Ed25519 public key, raw 32 bytes encoded as base64. The agent's "
            "DID is derived from this key on the server side."
        )
    )
    owner_operator_id: str = Field(
        description=(
            "Identifier of the operator that owns this agent — typically the "
            "tenant the agent belongs to."
        ),
        min_length=1,
        max_length=128,
    )


class _ResolveArgs(BaseModel):
    did: str = Field(
        description=(
            "The ``did:key:z...`` identifier to resolve. Returns the agent's "
            "public key, signature scheme, owner operator, and registration "
            "timestamp."
        )
    )


class IdentityRegisterTool(ColberToolBase):
    """Register a new agent on the Colber identity service."""

    service_name: ClassVar[str] = "identity"

    name: str = "colber_identity_register"
    description: str = (
        "Register a new agent on Colber by providing its Ed25519 public key. "
        "Returns the freshly-minted ``did:key`` and an internal agent id. "
        "Use this once per agent, before the agent participates in "
        "negotiations or builds a reputation."
    )
    args_schema: type[BaseModel] = _RegisterArgs

    def _call_colber(self, **kwargs: Any) -> Any:
        return self._client.identity.register(
            public_key=str(kwargs["public_key"]),
            owner_operator_id=str(kwargs["owner_operator_id"]),
        )


class IdentityResolveTool(ColberToolBase):
    """Resolve a Colber DID to its full identity record."""

    service_name: ClassVar[str] = "identity"

    name: str = "colber_identity_resolve"
    description: str = (
        "Look up an agent's identity record by its ``did:key`` identifier. "
        "Returns the agent's public key (for signature verification), owner "
        "operator id, and registration / revocation timestamps. Use this "
        "before trusting a counterparty's signed payload."
    )
    args_schema: type[BaseModel] = _ResolveArgs

    def _call_colber(self, **kwargs: Any) -> Any:
        return self._client.identity.resolve(str(kwargs["did"]))


__all__ = ["IdentityRegisterTool", "IdentityResolveTool"]
